iavl/iavl_proof2.go

```diff
@@ -167,52 +171,265 @@ func (proof *KeyNotExistsProof) Verify(key []byte, root []byte) error {
+			return errors.Errorf("paths #%d and #%d are not adjacent", i, i+1)
+		}
+	}
+	return nil
 }
 
-func (proof *KeyRangeExistsProof) Verify(key []byte, value []byte, root []byte) bool {
-	return false
+// Verify that a range proof is valid.
+func (proof *KeyRangeProof) Verify(
+	startKey, endKey []byte, limit int, keys, values [][]byte, root []byte,
+) error {
+	if len(proof.PathToKeys) != len(keys) || len(values) != len(keys) {
+		return errors.New("wrong number of keys or values for proof")
+	}
+
+	// If we've used a limit, our end key can be much greater than the right
+	// node key, ex: with the query start=0, end=FF, limit=10, the returned
+	// key range might be from 12 to D4, and the right node key will be the
+	// one following D4, say DF. In this case, we set our start and end keys
+	// to the key range returned.
+	if len(keys) == limit && limit > 0 {
+		endKey = keys[len(keys)-1]
+		startKey = keys[0]
+	}
+
+	// If startKey > endKey, reverse the keys and values, since our proofs are
+	// always in ascending order.
+	if bytes.Compare(startKey, endKey) == 1 {
+		startKey, endKey = endKey, startKey
+
+		ks, vs := [][]byte{}, [][]byte{}
+		for i := len(keys) - 1; i >= 0; i-- {
+			ks = append(ks, keys[i])
+			vs = append(vs, values[i])
+		}
+		keys, values = ks, vs
+	}
+
+	if proof.LeftPath != nil {
+		if err := proof.LeftPath.verify(proof.LeftNode, root); err != nil {
+			return errors.Wrap(err, "failed to verify left path")
+		}
+		if bytes.Compare(proof.LeftNode.KeyBytes, startKey) != -1 {
+			return errors.New("left node must be lesser than start key")
+		}
+	}
+	if proof.RightPath != nil {
+		if err := proof.RightPath.verify(proof.RightNode, root); err != nil {
+			return errors.Wrap(err, "failed to verify right path")
+		}
+		if bytes.Compare(endKey, proof.RightNode.KeyBytes) != -1 {
+			return errors.New("right node must be greater than end key")
+		}
+	}
+
+	// Verify that all paths are adjacent to one another.
+	if err := proof.verifyPathAdjacency(); err != nil {
+		return err
+	}
+
+	// If proof.PathToKeys is empty, it means we have an empty range. This range
+	// can be between keys, or outside of the range of existing keys.
+	if len(proof.PathToKeys) == 0 {
+		if proof.LeftPath == nil && proof.RightPath == nil {
+			return errors.New("proof is incomplete")
+		}
+		if proof.LeftPath == nil && proof.RightPath != nil {
+			// Range is outisde and to the left of existing keys.
+			if !proof.RightPath.isLeftmost() {
+				return errors.New("right path is not leftmost")
+			}
+		} else if proof.RightPath == nil && proof.LeftPath != nil {
+			// Range is outisde and to the right of existing keys.
+			if !proof.LeftPath.isRightmost() {
+				return errors.New("left path is not rightmost")
+			}
+		}
+	} else {
+		// If we've reached this point, it means our range isn't empty, and we have
+		// a list of keys.
+		for i, path := range proof.PathToKeys {
+			leafNode := IAVLProofLeafNode{KeyBytes: keys[i], ValueBytes: values[i]}
```

Comment:
should we make sure the keys are ordered earlier? or just leave it to here (since we already check that the paths are ordered, this should be sufficient)